### PR TITLE
Authorization by username and password

### PR DESCRIPTION
**Pull request by Anonym-tsk**
## Description:
Support authorization by username and password

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#296

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/__main__.py [hunk 1]
```diff
@@ -474,7 +474,11 @@ def parse_args(argv):
                                       help="Create a simple web server for a dashboard.")
     dashboard.add_argument("--port", help="The HTTP port to open connections on. Defaults to 6052.",
                            type=int, default=6052)
-    dashboard.add_argument("--password", help="The optional password to require for all requests.",
+    dashboard.add_argument("--username", help="The optional username to require "
+                                              "for authentication.",
+                           type=str, default='')
```
Comment (OttoWinter) [inline esphome/__main__.py]: Why should we have a username when there's only one possible user? Usernames are for cases where there are multiple possible users.

This also has some other issues:
 - Existing user's setups will break, as now they're asked for a username too even though they only supplied a password in their config.
 - We should validate that both username and password is given.

Please change it back to only accept password and make the login form conditional (only display the username input if on Hass.io)

File: esphome/dashboard/dashboard.py [hunk 2]
```diff
@@ -530,9 +533,12 @@ class EditRequestHandler(BaseHandler):
...
     def get(self, configuration=None):
-        # pylint: disable=no-value-for-parameter
-        with open(settings.rel_path(configuration), 'r') as f:
-            content = f.read()
+        try:
+            # pylint: disable=no-value-for-parameter
+            with open(settings.rel_path(configuration), 'r') as f:
+                content = f.read()
+        except IOError:
+            content = ''
```
Comment (OttoWinter) [inline esphome/dashboard/dashboard.py]: I don't think we should return a blank config if something with IO goes wrong.

What is this for? Returning no content if the file *does not exist* ? If so, please make it explicit by checking if the file exists first and then returning the empty string.

File: esphome/dashboard/static/esphome.css [hunk 3]
```diff
@@ -47,6 +47,7 @@ i.very-large {
 }
 
 .log {
+  height: 100%;
```
Comment (OttoWinter) [inline esphome/dashboard/static/esphome.css]: What is this change for? Please split up PRs if you want to add multiple features/fixes.

File: esphome/dashboard/static/esphome.js [hunk 4]
```diff
@@ -574,6 +574,7 @@ const editModalElem = document.getElementById("modal-editor");
 const editorElem = editModalElem.querySelector("#editor");
 const editor = ace.edit(editorElem);
 let activeEditorConfig = null;
+let activeEditorSecrets = false;
```
Comment (OttoWinter) [inline esphome/dashboard/static/esphome.js]: Is this for editing the `secrets.yaml` file? If so, please split that up into another PR.

File: esphome/dashboard/dashboard.py (hunk 2 above)
Comment (Anonym-tsk) [reply esphome/dashboard/dashboard.py]: Unfortunately, Py2 does'n have `FileNotFoundError` and it's normal way to check `IOError`. Now if you try open file that doesn't exists, you get html code in editor
```
<html><title>500: Internal Server Error</title><body>500: Internal Server Error</body></html>
```
I think, it's better, to return empty string.

**Comment by Anonym-tsk** [comment]
I splitted this PR:
1. Secrets editor: #672

File: esphome/dashboard/dashboard.py (hunk 2 above)
Comment (OttoWinter) [reply esphome/dashboard/dashboard.py]: Yes I don't have anything against returning an empty string if the file does not exist.

But the except case should not catch *all* IOErrors. There might be ones where we don't want to send an empty string. (Think what if the read failed for another reason, and then user presses save, the contents of the file will be cleared)

A quick check for existance before opening the file is fine, catching all IOErrors is not.

Comment (Anonym-tsk) [reply esphome/dashboard/dashboard.py]: https://github.com/esphome/esphome/pull/672/files#diff-19785080f51a55a345499b1f24d17824R533

File: esphome/__main__.py (hunk 1 above)
Comment (Anonym-tsk) [reply esphome/__main__.py]: > Why should we have a username when there's only one possible user?

For safety. Username + password are safer than only password.

> Existing user's setups will break, as now they're asked for a username too even though they only supplied a password in their config

No, it's not breaking change. If you don't configure username, you can use only password, as previously.

> We should validate that both username and password is given

No, it's possible to use only password.

**Review by OttoWinter — APPROVED** [review]
